### ISSUE TITLE
Lost minutes and seconds @Field(type = FieldType.Date, format = DateFormat.date_optional_time)

### ISSUE THREAD

**Issue by lihengcn** [spring-projects/spring-data-elasticsearch#2308]
spring-boot-starter-data-elasticsearch 2.6.11
elasticsearch 7.17.6

```
@Data
@NoArgsConstructor
@AllArgsConstructor
@ToString
@Builder
@Document(indexName = "shopping")
public class Product {

    @Id
    private String id;

    @Field(type = FieldType.Text)
    private String title;

    @Field(type = FieldType.Keyword)
    private String category;

    @Field(type = FieldType.Double)
    private Double price;

    @Field(type = FieldType.Keyword, index = false)
    private String images;

    @Field(type = FieldType.Date, format = DateFormat.date_optional_time)
    private LocalDateTime createTime;

}
```

```
@Test
public void create() {
    Product product = Product.builder().title("phone").category("iphone").price(2399.00).images("http://www.baidu.com/1.jpg").createTime(LocalDateTime.now()).build();
    productRepository.save(product);
}
```
```
{
  "_index" : "shopping",
  "_type" : "_doc",
  "_id" : "T5Kra4MBdC5CrdATYeqa",
  "_version" : 1,
  "_seq_no" : 0,
  "_primary_term" : 1,
  "found" : true,
  "_source" : {
    "_class" : "com.example.elasticsearch.po.Product",
    "title" : "phone",
    "category" : "iphone",
    "price" : 2399.0,
    "images" : "http://www.baidu.com/1.jpg",
    "createTime" : "2022-09-24"
  }
}
```
Minutes and seconds are lost after saving

**Comment by lihengcn** [comment]
When I switched to version spring-boot-starter-data-elasticsearch 2.4.5, the minutes and seconds were displayed, I don't know what the reason is

**Comment by sothawo** [comment]
Since Spring Data Elasticsearch 4.2 the `ElasticsearchDateConverter` that is used to convert temporal data uses the 
`java.time.format.DateTimeFormatter` for date formats that are defined by a date pattern. And every value of `org.
springframework.data.elasticsearch.annotations.DateFormat` uses such a pattern, you can see them [in the code](https://github.com/spring-projects/spring-data-elasticsearch/blob/main/src/main/java/org/springframework/data/elasticsearch/annotations/DateFormat.java#L30-L89).

The pattern that is used for `date_optional_time` id `"uuuu-MM-dd['T'HH:mm:ss.SSSXXX]"`. Notice that the optional 
time part has a time zone definition is this pattern ("XXX"). The Elasticsearch definition for this also has the 
time zone in it [see the docs](https://www.elastic.co/guide/en/elasticsearch/reference/8.4/mapping-date-format.html#built-in-date-formats).

The problem you are facing now, is that you use a `java.time.LocalDateTime` which has no time zone information. And 
when formatting this the time is ignored:

```java
public static void main(String[] args) {
	var patternOptionalDateTime = "uuuu-MM-dd['T'HH:mm:ss.SSSXXX]";
	var formatter = DateTimeFormatter.ofPattern(patternOptionalDateTime);
	var formatted = formatter.format(LocalDateTime.now());
	System.out.println("formatted: " + formatted);
	var parsed = formatter.parse(formatted);
	System.out.println("parsed: " + parsed.toString());
	}
```

```java
formatted: 2022-09-24
parsed: {},ISO resolved to 2022-09-24
```

You have 2 options now: You can either switch to use a ZonedDateTime:

```java
public static void main(String[] args) {
	var patternOptionalDateTime = "uuuu-MM-dd['T'HH:mm:ss.SSSXXX]";
	var formatter = DateTimeFormatter.ofPattern(patternOptionalDateTime);
	var formatted = formatter.format(ZonedDateTime.now());
	System.out.println("formatted: " + formatted);
	var parsed = formatter.parse(formatted);
	System.out.println("parsed: " + parsed.toString());
	}
```

```java
formatted: 2022-09-24T11:35:55.526+02:00
	parsed: {OffsetSeconds=7200, InstantSeconds=1664012155},ISO resolved to 2022-09-24T11:35:55.526
```

Or you can use a custom formatter pattern that does not include the time zone:

```java
public static void main(String[] args) {
	var patternOptionalDateTime = "uuuu-MM-dd['T'HH:mm:ss.SSS]";
	var formatter = DateTimeFormatter.ofPattern(patternOptionalDateTime);
	var formatted = formatter.format(LocalDateTime.now());
	System.out.println("formatted: " + formatted);
	var parsed = formatter.parse(formatted);
	System.out.println("parsed: " + parsed.toString());
}
```
```
formatted: 2022-09-24T11:37:10.498
parsed: {},ISO resolved to 2022-09-24T11:37:10.498
```

For the use in Spring Data Elasticsearch the property definition would need to change to:
```java
@Field(type = FieldType.Date, format = {}, pattern = "uuuu-MM-dd['T'HH:mm:ss.SSS]")
private LocalDateTime createTime;
```
Why did it work in older versions? We were using code from the Elasticsearch core libraries then which we cannot do 
anymore after the license change. This code handled date formats in a much more flexible way.

**Comment by lihengcn** [comment]
Great, thanks for your information!
